File: python3/find_players_with_zero_or_one_losses.py
"""
Space   : O(n)
Time    : O(n log n)
"""

class Solution:
    def findWinners(self, matches: List[List[int]]) -> List[List[int]]:
        ans = [[], []]
        lose = {}
        players = set()

        for w, l in matches:
            if l in lose:
                lose[l] += 1
            else:
                lose[l] = 1

            players.add(w)
            players.add(l)

        players = sorted(list(players))
        for p in players:
            if p not in lose:
                ans[0].append(p)
            elif lose[p] == 1:
                ans[1].append(p)

        return ans

File: golang/find_players_with_zero_or_one_losses.go
package main

import "sort"

/*
Space	: O(n)
Time	: O(n)
*/

func findWinners(matches [][]int) [][]int {
	ans := [][]int{[]int{}, []int{}}
	winnerMap := map[int][]int{}
	loserMap := map[int][]int{}
	participants := map[int]bool{}

	for _, match := range matches {
		winner := match[0]
		loser := match[1]

		winnerMap[winner] = append(winnerMap[winner], loser)
		loserMap[loser] = append(loserMap[loser], winner)
		participants[winner] = true
		participants[loser] = true
	}

	for k, _ := range participants {
		_, isWinner := winnerMap[k]
		losingMatches, isLoser := loserMap[k]

		if isWinner && !isLoser {
			ans[0] = append(ans[0], k)
		} else if isLoser && len(losingMatches) == 1 {
			ans[1] = append(ans[1], k)
		}
	}

	sort.Ints(ans[0])
	sort.Ints(ans[1])

	return ans
}
